File: src/lib/desktop/accounting.ts
import { invoke } from '@tauri-apps/api/core';
import type { TauriActionResult } from '@/lib/desktop/admin-types';
import { isTauriBackendAvailable } from '@/lib/desktop/runtime';

function requireTauri(): void {
  if (!isTauriBackendAvailable()) {
    throw new Error('Tauri accounting API requires the desktop shell');
  }
}

export async function tauriAccountingGetSettings(orgId?: string) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>>>('accounting_get_settings_cmd', { orgId: orgId ?? null });
}

export async function tauriAccountingUpdateSettings(orgId: string, patch: Record<string, unknown>) {
  requireTauri();
  return invoke<TauriActionResult<boolean>>('accounting_update_settings_cmd', { orgId, patch });
}

export async function tauriAccountingGetChartOfAccounts(orgId?: string) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>[]>>('accounting_get_chart_of_accounts_cmd', {
    orgId: orgId ?? null,
  });
}

export async function tauriAccountingGetJournalEntries(orgId?: string, limit?: number) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>[]>>('accounting_get_journal_entries_cmd', {
    orgId: orgId ?? null,
    limit: limit ?? 100,
  });
}

export async function tauriAccountingGetTrialBalance(orgId?: string) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>[]>>('accounting_get_trial_balance_cmd', {
    orgId: orgId ?? null,
  });
}

export async function tauriAccountingGetStatements(
  orgId?: string,
  range?: { startDate?: string; endDate?: string }
) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>>>('accounting_get_statements_cmd', {
    orgId: orgId ?? null,
    range: range ?? null,
  });
}

export async function tauriAccountingBackfill(orgId?: string) {
  requireTauri();
  return invoke<TauriActionResult<{ posted: number }>>('accounting_backfill_cmd', { orgId: orgId ?? null });
}

export async function tauriAccountingReconcile(
  orgId: string,
  countedCash: number,
  countedBank: number,
  notes?: string
) {
  requireTauri();
  return invoke<TauriActionResult<Record<string, unknown>>>('accounting_reconcile_cmd', {
    orgId,
    countedCash,
    countedBank,
    notes: notes ?? null,
  });
}

File: src-tauri/src/commands/accounting.rs
use serde::Deserialize;
use serde_json::Value;
use tauri::State;

use crate::accounting::service::{
    backfill, get_chart_of_accounts, get_journal_entries, get_settings, get_statements,
    get_trial_balance, reconcile, update_settings,
};
use crate::admin::service::ActionResult;
use crate::auth::session::SessionState;
use crate::state::AppState;

#[derive(Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct DateRangeInput {
    pub start_date: Option<String>,
    pub end_date: Option<String>,
}

#[tauri::command]
pub fn accounting_get_settings_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    org_id: Option<String>,
) -> ActionResult<Value> {
    get_settings(&app, &session, org_id.as_deref())
}

#[tauri::command]
pub fn accounting_update_settings_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    org_id: String,
    patch: Value,
) -> ActionResult<bool> {
    update_settings(&app, &session, &org_id, patch)
}

#[tauri::command]
pub fn accounting_get_chart_of_accounts_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    org_id: Option<String>,
) -> ActionResult<Vec<Value>> {
    get_chart_of_accounts(&app, &session, org_id.as_deref())
}

#[tauri::command]
pub fn accounting_get_journal_entries_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    org_id: Option<String>,
    limit: Option<usize>,
) -> ActionResult<Vec<Value>> {
    get_journal_entries(&app, &session, org_id.as_deref(), limit.unwrap_or(100))
}

#[tauri::command]
pub fn accounting_get_trial_balance_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    org_id: Option<String>,
) -> ActionResult<Vec<Value>> {
    get_trial_balance(&app, &session, org_id.as_deref())
}

#[tauri::command]
pub fn accounting_get_statements_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    org_id: Option<String>,
    _range: Option<DateRangeInput>,
) -> ActionResult<Value> {
    get_statements(&app, &session, org_id.as_deref())
}

#[tauri::command]
pub fn accounting_backfill_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    org_id: Option<String>,
) -> ActionResult<Value> {
    backfill(&app, &session, org_id.as_deref())
}

#[tauri::command]
pub fn accounting_reconcile_cmd(
    app: State<'_, AppState>,
    session: State<'_, SessionState>,
    org_id: String,
    counted_cash: f64,
    counted_bank: f64,
    notes: Option<String>,
) -> ActionResult<Value> {
    reconcile(
        &app,
        &session,
        &org_id,
        counted_cash,
        counted_bank,
        notes.as_deref(),
    )
}
